### DISCUSSION TITLE
ImageMagick Magick.NET to Avalonia WriteableBitmap

### DISCUSSION THREAD

**Discussion by Ruben2776** [AvaloniaUI/Avalonia#17365] [Q&A]
I want to convert a MagickImage to a WriteableBitmap or Bitmap, since Magick.net has support for .avif, .jxl, RAW Camera photos, photoshop etc. 
For WPF, this could be achieved with the extension method [ToBitmapSource](https://github.com/dlemstra/Magick.NET/blob/main/src/Magick.NET.SystemWindowsMedia/IMagickImageExtentions.cs). 
I'm currently achieving it by setting the format as `Png` and writing it to a `MemorySteam`, but that is inefficient. I'm looking for a similar way of correctly converting the pixel data, to use for an Avalonia image.

**Comment by stevemonaco** [comment]
Assuming you have an uncompressed pixel buffer, your main approaches are:
1. Use `WriteableBitmap`. You can assign this to an `Image.Source`. Lock it and use `Marshal.Copy` or pointers. eg. https://github.com/AvaloniaUI/Avalonia/blob/68a626f678b7fa4b1d1497cf7f419f7aebf703ee/samples/RenderDemo/Pages/WriteableBitmapPage.cs#L41 and https://github.com/stevemonaco/AvaloniaDemos/tree/master/RealTimeBitmapAdapter
2. Use `SKBitmap.InstallPixels` and SkiaSharp rendering instead of dealing with the `Image` control. eg. https://github.com/stevemonaco/AvaloniaDemos/tree/master/SkiaBitmapAdapter

You'll need to transcode the raw pixel data if the pixel formats are different.

**Comment by thevortexcloud** [comment]
One other approach not mentioned is you can actually implement your own custom bitmap. The interface is public.

https://github.com/AvaloniaUI/Avalonia/blob/68a626f678b7fa4b1d1497cf7f419f7aebf703ee/src/Avalonia.Base/Media/IImage.cs#L5-L8

There is an example here that wraps a Skia bitmap:

https://github.com/AvaloniaUI/Avalonia/discussions/13610#discussioncomment-7576208
